std-bits/table/src/main/java/org/enso/table/parsing/NumberParser.java

```diff
@@ -0,0 +1,279 @@
+    }
+
+    private static Pattern getPattern(boolean allowDecimal, boolean allowCurrency, boolean allowScientific, boolean trimValues, int patternIndex, String separators) {
+        if (allowScientific && !allowDecimal) {
+            throw new IllegalArgumentException("Scientific notation requires decimal numbers.");
+        }
+
+        if (patternIndex >= (allowCurrency ? 6 : 2)) {
+            return null;
+        }
+
+        var INTEGER = "(?<integer>(\\d*)" + (separators.length() == 1 ? "" : "|(\\d{1,3}([" + separators.charAt(0) + "]\\d{3})*)") + ")";
+
+        var decimalPoint = (separators.length() == 1 ? separators : separators.charAt(1));
+        var NUMBER = INTEGER + (allowDecimal ? "(?<decimal>[" + decimalPoint + "]\\d*)?" : "") + (allowScientific ? EXP : "");
+
+        var pattern = switch (patternIndex) {
+            case 0 -> SIGN + NUMBER;
+            case 1 -> BRACKETS + NUMBER + BRACKET_CLOSE;
+            case 2 -> SIGN + CCY + SPACE + NUMBER;
+            case 3 -> CCY + SPACE + SIGN + NUMBER;
+            case 4 -> SIGN + NUMBER + CCY;
+            case 5 -> BRACKETS + CCY + SPACE + NUMBER + BRACKET_CLOSE;
+            default -> throw new IllegalArgumentException("Invalid pattern index: " + patternIndex);
+        };
+
+        if (trimValues) {
+            pattern = SPACE + pattern + SPACE;
+        }
+
+        return PATTERNS.computeIfAbsent("^" + pattern + "$", Pattern::compile);
+    }
+
+    private final boolean allowDecimal;
+    private final boolean allowCurrency;
+    private final boolean allowLeadingZeros;
+    private final boolean allowScientific;
+    private final boolean trimValues;
+    private final String separators;
+
+    /**
+     * Creates a new integer instance of this parser.
+     *
+     * @param allowCurrency whether to allow currency symbols
+     * @param allowLeadingZeros whether to allow leading zeros
+     * @param trimValues whether to trim the input values
+     * @param thousandSeparator the thousand separator to use
+     */
+    public NumberParser(boolean allowCurrency, boolean allowLeadingZeros, boolean trimValues, String thousandSeparator) {
+        this.allowCurrency = allowCurrency;
+        this.allowLeadingZeros = allowLeadingZeros;
+        this.trimValues = trimValues;
+
+        this.allowDecimal = false;
+        this.allowScientific = false;
+
+        this.separators = thousandSeparator == null ? null : (thousandSeparator + '_');
+    }
+
+    /**
+     * Creates a new decimal instance of this parser.
+     *
+     * @param allowCurrency whether to allow currency symbols
+     * @param allowLeadingZeros whether to allow leading zeros
+     * @param trimValues whether to trim the input values
+     * @param allowScientific whether to allow scientific notation
+     */
+    public NumberParser(boolean allowCurrency, boolean allowLeadingZeros, boolean trimValues, boolean allowScientific) {
```

Comment:
It is hard to distinguish which constructor is making an integer parser and which one is making a decimal one, it's just by the type of one of the arguments.

I think for code readability it would be good to make these constructors private and then expose two static factory methods: `createIntegerParser` and `createDecimalParser` (possibly with the 2 overrides). That way at usage site it will be clear which variant is used.